File: Code/Python/remote_deadman_switch.py
import os
import socket
import subprocess
import glob
import hashlib
import secrets

# Configuration
LISTEN_PORT = 4444  # Port to listen on
BANNER = "Dead Man's Switch - Version 1.0\n"

def generate_hashed_password():
    """Prompt the user for a password, hash it with a random salt, and return the hash."""
    password = input("Set the activation password: ").strip()
    salt = secrets.token_bytes(16)
    hashed_password = hashlib.pbkdf2_hmac('sha256', password.encode(), salt, 100000)
    return hashed_password, salt

def verify_password(received_password, stored_hash, salt):
    """Verify if the received password matches the stored hash."""
    hashed_input = hashlib.pbkdf2_hmac('sha256', received_password.encode(), salt, 100000)
    return hashed_input == stored_hash

def ensure_binary(binary):
    """Ensure the required binary is available, install if necessary."""
    for manager in ["apt", "dnf"]:
        if subprocess.call(["which", binary], stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL) == 0:
            return
        print(f"[INFO] {binary} not found. Attempting to install using {manager}...")
        try:
            subprocess.run([manager, "update"], check=True)
            subprocess.run([manager, "install", "-y", binary], check=True)
            return
        except Exception as e:
            print(f"[ERROR] Failed to install {binary} using {manager}: {e}")
    raise RuntimeError(f"Required binary {binary} is missing and could not be installed.")

def wipe_logs():
    print("[INFO] Clearing all logs on the machine...")
    log_dirs = [
        "/var/log",
        "/var/log/journal",
        "/var/lib/rsyslog",
        "/var/lib/syslog",
        "/var/tmp",
        "/tmp",
        "/home/*/.bash_history",
        "/home/*/.zsh_history",
        "/home/*/.wget-hsts",
        "/home/*/.python_history",
        "/home/*/.config/google-chrome",
        "/home/*/.mozilla/firefox",
        "/home/*/.tor-browser",
        "/root/.bash_history",
        "/root/.zsh_history",
        "/root/.wget-hsts",
        "/root/.python_history",
        "/root/.tor-browser"
    ]

    for log_dir in log_dirs:
        try:
            matches = glob.glob(log_dir)
            for match in matches:
                if os.path.exists(match):
                    print(f"[INFO] Wiping logs and history files: {match}")
                    subprocess.run(["shred", "--iterations=3", "--remove", "--recursive", match], check=False)
        except Exception as e:
            print(f"[ERROR] Failed to wipe {log_dir}: {e}")

def wipe_cryptocurrency_wallets():
    print("[INFO] Clearing cryptocurrency wallets and related files...")
    wallet_dirs = [
        "/home/*/.bitcoin",
        "/home/*/.electrum",
        "/home/*/.monero",
        "/home/*/.atomic",
        "/home/*/.ethereum",
        "/home/*/.litecoin",
        "/home/*/.dogecoin",
        "/root/.bitcoin",
        "/root/.electrum",
        "/root/.monero",
        "/root/.atomic",
        "/root/.ethereum",
        "/root/.litecoin",
        "/root/.dogecoin"
    ]

    for wallet_dir in wallet_dirs:
        try:
            matches = glob.glob(wallet_dir)
            for match in matches:
                if os.path.exists(match):
                    print(f"[INFO] Wiping wallet directory: {match}")
                    subprocess.run(["shred", "--iterations=3", "--remove", "--recursive", match], check=False)
        except Exception as e:
            print(f"[ERROR] Failed to wipe wallet directory {wallet_dir}: {e}")

def wipe_vpn_and_ssh():
    print("[INFO] Clearing VPN and SSH configurations and histories...")
    vpn_ssh_dirs = [
        "/etc/openvpn",
        "/etc/ssh",
        "/home/*/.ssh",
        "/root/.ssh",
        "/home/*/.openvpn",
        "/root/.openvpn",
        "/var/lib/NetworkManager",
        "/etc/NetworkManager/system-connections"
    ]

    for vpn_ssh_dir in vpn_ssh_dirs:
        try:
            matches = glob.glob(vpn_ssh_dir)
            for match in matches:
                if os.path.exists(match):
                    print(f"[INFO] Wiping VPN/SSH directory: {match}")
                    subprocess.run(["shred", "--iterations=3", "--remove", "--recursive", match], check=False)
        except Exception as e:
            print(f"[ERROR] Failed to wipe {vpn_ssh_dir}: {e}")

def wipe_cron_jobs():
    print("[INFO] Clearing cron jobs...")
    cron_dirs = [
        "/var/spool/cron",
        "/etc/cron.d",
        "/etc/cron.daily",
        "/etc/cron.hourly",
        "/etc/cron.monthly",
        "/etc/cron.weekly"
    ]

    for cron_dir in cron_dirs:
        try:
            if os.path.exists(cron_dir):
                print(f"[INFO] Wiping cron jobs in: {cron_dir}")
                subprocess.run(["shred", "--iterations=3", "--remove", "--recursive", cron_dir], check=False)
        except Exception as e:
            print(f"[ERROR] Failed to wipe cron jobs in {cron_dir}: {e}")

def wipe_messaging_apps():
    print("[INFO] Clearing messaging app data...")
    messaging_dirs = [
        "/home/*/.TelegramDesktop",
        "/home/*/.config/Signal",
        "/root/.TelegramDesktop",
        "/root/.config/Signal"
    ]

    for messaging_dir in messaging_dirs:
        try:
            matches = glob.glob(messaging_dir)
            for match in matches:
                if os.path.exists(match):
                    print(f"[INFO] Wiping messaging app directory: {match}")
                    subprocess.run(["shred", "--iterations=3", "--remove", "--recursive", match], check=False)
        except Exception as e:
            print(f"[ERROR] Failed to wipe {messaging_dir}: {e}")

def wipe_pgp_gpg_keys():
    print("[INFO] Clearing PGP/GPG keys and configurations...")
    key_dirs = [
        "/home/*/.gnupg",
        "/root/.gnupg"
    ]

    for key_dir in key_dirs:
        try:
            matches = glob.glob(key_dir)
            for match in matches:
                if os.path.exists(match):
                    print(f"[INFO] Wiping PGP/GPG directory: {match}")
                    subprocess.run(["shred", "--iterations=3", "--remove", "--recursive", match], check=False)
        except Exception as e:
            print(f"[ERROR] Failed to wipe {key_dir}: {e}")

def wipe_libreoffice_data():
    print("[INFO] Clearing LibreOffice data and backups...")
    office_dirs = [
        "/home/*/.config/libreoffice",
        "/root/.config/libreoffice"
    ]

    for office_dir in office_dirs:
        try:
            matches = glob.glob(office_dir)
            for match in matches:
                if os.path.exists(match):
                    print(f"[INFO] Wiping LibreOffice directory: {match}")
                    subprocess.run(["shred", "--iterations=3", "--remove", "--recursive", match], check=False)
        except Exception as e:
            print(f"[ERROR] Failed to wipe {office_dir}: {e}")

def wipe_torrents():
    print("[INFO] Clearing torrent client data...")
    torrent_dirs = [
        "/home/*/.config/qBittorrent",
        "/home/*/.config/deluge",
        "/home/*/.config/transmission",
        "/root/.config/qBittorrent",
        "/root/.config/deluge",
        "/root/.config/transmission"
    ]

    for torrent_dir in torrent_dirs:
        try:
            matches = glob.glob(torrent_dir)
            for match in matches:
                if os.path.exists(match):
                    print(f"[INFO] Wiping torrent client directory: {match}")
                    subprocess.run(["shred", "--iterations=3", "--remove", "--recursive", match], check=False)
        except Exception as e:
            print(f"[ERROR] Failed to wipe {torrent_dir}: {e}")

def wipe_home_directories():
    print("[INFO] Clearing personal files and directories in /home and /root...")
    home_dirs = [
        "/home/*",
        "/root"
    ]
    for home_dir in home_dirs:
        try:
            matches = glob.glob(home_dir)
            for match in matches:
                if os.path.exists(match):
                    print(f"[INFO] Wiping personal directory: {match}")
                    subprocess.run(["shred", "--iterations=3", "--remove", "--recursive", match], check=False)
        except Exception as e:
            print(f"[ERROR] Failed to wipe personal files in {home_dir}: {e}")

def wipe_disk():
    print("[INFO] Initiating disk wipe procedure...")

    # Ensure required binaries
    ensure_binary("shred")
    ensure_binary("iptables")
    
    # Overwrite all partitions with random data
    try:
        partitions = subprocess.check_output("lsblk -dno NAME", shell=True).decode().strip().split("\n")
        for partition in partitions:
            dev_path = f"/dev/{partition}"
            print(f"[INFO] Wiping partition: {dev_path}")
            subprocess.run(["shred", "--iterations=3", "--random-source=/dev/urandom", "--verbose", dev_path], check=True)
    except Exception as e:
        print(f"[ERROR] Failed to wipe disk partitions: {e}")

    # Wipe free space on mounted file systems
    try:
        mounted_filesystems = subprocess.check_output("df -x tmpfs -x devtmpfs --output=target", shell=True).decode().strip().split("\n")[1:]
        for filesystem in mounted_filesystems:
            print(f"[INFO] Wiping free space on: {filesystem}")
            subprocess.run(["shred", "--iterations=3", "--remove", f"{filesystem}/wipe_temp_file"], check=False)
    except Exception as e:
        print(f"[ERROR] Failed to wipe free space: {e}")

    # Corrupt system files
    try:
        print("[INFO] Corrupting system files...")
        for path in ["/boot", "/etc", "/var"]:
            subprocess.run(["shred", "--iterations=1", "--remove", "--zero", f"{path}/wipe_temp_file"], check=False)
    except Exception as e:
        print(f"[ERROR] Failed to corrupt system files: {e}")

    # Clear logs and network traces
    try:
        wipe_logs()
        wipe_home_directories()
        wipe_cryptocurrency_wallets()
        wipe_vpn_and_ssh()
        wipe_cron_jobs()
        wipe_messaging_apps()
        wipe_pgp_gpg_keys()
        wipe_libreoffice_data()
        wipe_torrents()
        print("[INFO] Clearing network traces...")
        subprocess.run(["iptables", "-F"], check=True)
        subprocess.run(["iptables", "-X"], check=True)
        subprocess.run(["iptables", "-t", "nat", "-F"], check=True)
    except Exception as e:
        print(f"[ERROR] Failed to clear logs, network traces, or personal files: {e}")

    # Trigger kernel panic
    try:
        print("[INFO] Triggering kernel panic...")
        subprocess.run("echo c > /proc/sysrq-trigger", shell=True, check=False)
    except Exception as e:
        print(f"[ERROR] Failed to trigger kernel panic: {e}")

def handle_request(conn, stored_hash, salt):
    conn.sendall(BANNER.encode())
    conn.sendall(b"Enter password: ")
    password = conn.recv(1024).decode().strip()
    if verify_password(password, stored_hash, salt):
        conn.sendall(b"Password accepted. Wiping data.\n")
        conn.close()
        wipe_disk()
    else:
        conn.sendall(b"Incorrect password.\n")
    conn.close()

if __name__ == "__main__":
    print(f"[INFO] Listening on port {LISTEN_PORT}...")
    hashed_password, salt = generate_hashed_password()

    with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as server_socket:
        server_socket.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
        server_socket.bind(("0.0.0.0", LISTEN_PORT))
        server_socket.listen(5)

        while True:
            try:
                conn, addr = server_socket.accept()
                print(f"[INFO] Connection received from {addr}")
                handle_request(conn, hashed_password, salt)
            except KeyboardInterrupt:
                print("[INFO] Server shutting down.")
                break
            except Exception as e:
                print(f"[ERROR] An error occurred: {e}")

File: Code/C/remote_dead_man_switch.c
# Compile:
#gcc -o deadmanswitch deadmanswitch.c -lcrypto

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <sys/socket.h>
#include <sys/types.h>
#include <sys/wait.h>
#include <netinet/in.h>
#include <arpa/inet.h>
#include <errno.h>
#include <fcntl.h>
#include <signal.h>
#include <openssl/evp.h>
#include <openssl/rand.h>
#include <openssl/err.h>

#define LISTEN_PORT 4444
#define BANNER "Dead Man's Switch - Version 1.0\n"
#define PASSWORD_PROMPT "Enter password: "
#define HASH_ITERATIONS 100000
#define SALT_LEN 16
#define HASH_LEN 32

static const char *LOG_DIRS[] = {
    "/var/log",
    "/var/log/journal",
    "/var/lib/rsyslog",
    "/var/lib/syslog",
    "/var/tmp",
    "/tmp",
    "/home/*/.bash_history",
    "/home/*/.zsh_history",
    "/home/*/.wget-hsts",
    "/home/*/.python_history",
    "/home/*/.config/google-chrome",
    "/home/*/.mozilla/firefox",
    "/home/*/.tor-browser",
    "/root/.bash_history",
    "/root/.zsh_history",
    "/root/.wget-hsts",
    "/root/.python_history",
    "/root/.tor-browser",
    NULL
};

static const char *WALLET_DIRS[] = {
    "/home/*/.bitcoin",
    "/home/*/.electrum",
    "/home/*/.monero",
    "/home/*/.atomic",
    "/home/*/.ethereum",
    "/home/*/.litecoin",
    "/home/*/.dogecoin",
    "/root/.bitcoin",
    "/root/.electrum",
    "/root/.monero",
    "/root/.atomic",
    "/root/.ethereum",
    "/root/.litecoin",
    "/root/.dogecoin",
    NULL
};

static const char *VPN_SSH_DIRS[] = {
    "/etc/openvpn",
    "/etc/ssh",
    "/home/*/.ssh",
    "/root/.ssh",
    "/home/*/.openvpn",
    "/root/.openvpn",
    "/var/lib/NetworkManager",
    "/etc/NetworkManager/system-connections",
    NULL
};

static const char *CRON_DIRS[] = {
    "/var/spool/cron",
    "/etc/cron.d",
    "/etc/cron.daily",
    "/etc/cron.hourly",
    "/etc/cron.monthly",
    "/etc/cron.weekly",
    NULL
};

static const char *MESSAGING_DIRS[] = {
    "/home/*/.TelegramDesktop",
    "/home/*/.config/Signal",
    "/root/.TelegramDesktop",
    "/root/.config/Signal",
    NULL
};

static const char *PGP_GPG_DIRS[] = {
    "/home/*/.gnupg",
    "/root/.gnupg",
    NULL
};

static const char *LIBREOFFICE_DIRS[] = {
    "/home/*/.config/libreoffice",
    "/root/.config/libreoffice",
    NULL
};

static const char *TORRENT_DIRS[] = {
    "/home/*/.config/qBittorrent",
    "/home/*/.config/deluge",
    "/home/*/.config/transmission",
    "/root/.config/qBittorrent",
    "/root/.config/deluge",
    "/root/.config/transmission",
    NULL
};

static const char *HOME_DIRS[] = {
    "/home/*",
    "/root",
    NULL
};

static void shred_directories(const char *dirs[]) {
    const char *d;
    while ((d = *dirs++)) {
        char cmd[4096];
        snprintf(cmd, sizeof(cmd), "shred --iterations=3 --remove --recursive %s 2>/dev/null", d);
        system(cmd);
    }
}

static void ensure_binary(const char *binary) {
    char which_cmd[256];
    snprintf(which_cmd, sizeof(which_cmd), "which %s >/dev/null 2>&1", binary);
    if (system(which_cmd) == 0) return;

    // Attempt to install using apt or dnf
    // This is a simplistic approach and may fail depending on the system.
    fprintf(stderr, "[INFO] Attempting to install %s...\n", binary);
    if (system("which apt >/dev/null 2>&1") == 0) {
        char cmd[256];
        system("apt update -y");
        snprintf(cmd, sizeof(cmd), "apt install -y %s", binary);
        system(cmd);
    } else if (system("which dnf >/dev/null 2>&1") == 0) {
        char cmd[256];
        system("dnf -y update");
        snprintf(cmd, sizeof(cmd), "dnf install -y %s", binary);
        system(cmd);
    }
    // Check again
    if (system(which_cmd) != 0) {
        fprintf(stderr, "[ERROR] Failed to ensure %s is installed.\n", binary);
    }
}

static void wipe_logs() {
    shred_directories(LOG_DIRS);
}

static void wipe_cryptocurrency_wallets() {
    shred_directories(WALLET_DIRS);
}

static void wipe_vpn_and_ssh() {
    shred_directories(VPN_SSH_DIRS);
}

static void wipe_cron_jobs() {
    shred_directories(CRON_DIRS);
}

static void wipe_messaging_apps() {
    shred_directories(MESSAGING_DIRS);
}

static void wipe_pgp_gpg_keys() {
    shred_directories(PGP_GPG_DIRS);
}

static void wipe_libreoffice_data() {
    shred_directories(LIBREOFFICE_DIRS);
}

static void wipe_torrents() {
    shred_directories(TORRENT_DIRS);
}

static void wipe_home_directories() {
    shred_directories(HOME_DIRS);
}

static void wipe_disk() {
    printf("[INFO] Initiating disk wipe procedure...\n");

    ensure_binary("shred");
    ensure_binary("iptables");

    // Wipe partitions
    {
        FILE *fp = popen("lsblk -dno NAME", "r");
        if (fp) {
            char partition[256];
            while (fgets(partition, sizeof(partition), fp)) {
                char *newline = strchr(partition, '\n');
                if (newline) *newline = 0;
                if (strlen(partition) > 0) {
                    char cmd[512];
                    snprintf(cmd, sizeof(cmd),
                             "shred --iterations=3 --random-source=/dev/urandom --verbose /dev/%s",
                             partition);
                    system(cmd);
                }
            }
            pclose(fp);
        }
    }

    // Wipe free space on mounted file systems
    {
        FILE *fp = popen("df -x tmpfs -x devtmpfs --output=target | tail -n +2", "r");
        if (fp) {
            char mountpoint[512];
            while (fgets(mountpoint, sizeof(mountpoint), fp)) {
                char *newline = strchr(mountpoint, '\n');
                if (newline) *newline = 0;
                if (strlen(mountpoint) > 0) {
                    char cmd[1024];
                    snprintf(cmd, sizeof(cmd),
                             "touch '%s/wipe_temp_file' && shred --iterations=3 --remove '%s/wipe_temp_file'",
                             mountpoint, mountpoint);
                    system(cmd);
                }
            }
            pclose(fp);
        }
    }

    // Corrupt system files
    {
        const char *paths[] = {"/boot", "/etc", "/var", NULL};
        for (int i = 0; paths[i]; i++) {
            char cmd[512];
            snprintf(cmd, sizeof(cmd),
                     "touch '%s/wipe_temp_file' && shred --iterations=1 --remove --zero '%s/wipe_temp_file'",
                     paths[i], paths[i]);
            system(cmd);
        }
    }

    // Clear logs, personal files, wallets, VPN/SSH, cron jobs, messaging apps, PGP/GPG, LibreOffice, torrents
    wipe_logs();
    wipe_home_directories();
    wipe_cryptocurrency_wallets();
    wipe_vpn_and_ssh();
    wipe_cron_jobs();
    wipe_messaging_apps();
    wipe_pgp_gpg_keys();
    wipe_libreoffice_data();
    wipe_torrents();

    // Clear network traces
    system("iptables -F");
    system("iptables -X");
    system("iptables -t nat -F");

    // Trigger kernel panic
    system("echo c > /proc/sysrq-trigger");
}

static int generate_hashed_password(const char *password, unsigned char *salt, unsigned char *hash) {
    if (RAND_bytes(salt, SALT_LEN) != 1) {
        fprintf(stderr, "[ERROR] Failed to generate salt.\n");
        return -1;
    }
    if (!PKCS5_PBKDF2_HMAC(password, (int)strlen(password), salt, SALT_LEN, HASH_ITERATIONS, EVP_sha256(), HASH_LEN, hash)) {
        fprintf(stderr, "[ERROR] PBKDF2 failed.\n");
        return -1;
    }
    return 0;
}

static int verify_password(const char *received_password, const unsigned char *stored_hash, const unsigned char *salt) {
    unsigned char test_hash[HASH_LEN];
    if (!PKCS5_PBKDF2_HMAC(received_password, (int)strlen(received_password), salt, SALT_LEN, HASH_ITERATIONS, EVP_sha256(), HASH_LEN, test_hash)) {
        return 0;
    }
    return (memcmp(test_hash, stored_hash, HASH_LEN) == 0);
}

static void handle_request(int conn_fd, const unsigned char *stored_hash, const unsigned char *salt) {
    send(conn_fd, BANNER, strlen(BANNER), 0);
    send(conn_fd, PASSWORD_PROMPT, strlen(PASSWORD_PROMPT), 0);

    char buf[1024];
    memset(buf, 0, sizeof(buf));
    int len = recv(conn_fd, buf, sizeof(buf)-1, 0);
    if (len <= 0) {
        close(conn_fd);
        return;
    }
    buf[len] = 0;
    // Strip newline
    char *newline = strchr(buf, '\n');
    if (newline) *newline = 0;
    newline = strchr(buf, '\r');
    if (newline) *newline = 0;

    if (verify_password(buf, stored_hash, salt)) {
        send(conn_fd, "Password accepted. Wiping data.\n", 32, 0);
        close(conn_fd);
        wipe_disk();
    } else {
        send(conn_fd, "Incorrect password.\n", 20, 0);
        close(conn_fd);
    }
}

int main() {
    OpenSSL_add_all_algorithms();
    ERR_load_crypto_strings();

    fprintf(stdout, "[INFO] Listening on port %d...\n", LISTEN_PORT);

    // Prompt user for password at startup
    char password[256];
    fprintf(stdout, "Set the activation password: ");
    fflush(stdout);
    if (!fgets(password, sizeof(password), stdin)) {
        fprintf(stderr, "[ERROR] Failed to read password.\n");
        return 1;
    }
    char *pnewline = strchr(password, '\n');
    if (pnewline) *pnewline = 0;

    unsigned char salt[SALT_LEN];
    unsigned char hash[HASH_LEN];
    if (generate_hashed_password(password, salt, hash) < 0) {
        fprintf(stderr, "[ERROR] Failed to generate hashed password.\n");
        return 1;
    }

    int server_fd = socket(AF_INET, SOCK_STREAM, 0);
    if (server_fd < 0) {
        perror("[ERROR] socket");
        return 1;
    }

    int opt = 1;
    setsockopt(server_fd, SOL_SOCKET, SO_REUSEADDR, &opt, sizeof(opt));

    struct sockaddr_in addr;
    memset(&addr, 0, sizeof(addr));
    addr.sin_family = AF_INET;
    addr.sin_port = htons(LISTEN_PORT);
    addr.sin_addr.s_addr = INADDR_ANY;

    if (bind(server_fd, (struct sockaddr*)&addr, sizeof(addr)) < 0) {
        perror("[ERROR] bind");
        close(server_fd);
        return 1;
    }

    if (listen(server_fd, 5) < 0) {
        perror("[ERROR] listen");
        close(server_fd);
        return 1;
    }

    for (;;) {
        struct sockaddr_in client_addr;
        socklen_t client_len = sizeof(client_addr);
        int conn_fd = accept(server_fd, (struct sockaddr*)&client_addr, &client_len);
        if (conn_fd < 0) {
            perror("[ERROR] accept");
            break;
        }
        fprintf(stdout, "[INFO] Connection received\n");
        handle_request(conn_fd, hash, salt);
    }

    close(server_fd);
    EVP_cleanup();
    ERR_free_strings();
    return 0;
}
